### PR TITLE
Fixes #27 Add inline block parsing method

### LINKED ISSUE #27: Make it easier to block on parsing
To force parsing, one currently needs to do something like:

```
    val parser = ArgParser(args)
    val parsedArgs = ParsedArgs(parser)
    parser.force()
    // now you can use parsedArgs
```

It would be nice if the intermediate `ArgParser` didn't need to be named, so this could become a one-liner. Perhaps one could write something like:

```
    val parsedArgs = ArgParser(args).force(ParsedArgs)
```

This could be done by adding an optional parameter to force, or by creating a new method, hopefully with a better name.

### PR DESCRIPTION
**Pull request by shanethehat**

### CODE REVIEW

**Comment by shanethehat** [comment]
The code coverage report for this issue is reporting misses because Jacoco doesn't properly understand code generated from inlined methods, which is necessary in this case to use the reified type parameter. (https://stackoverflow.com/questions/39817814/no-coverage-report-for-inlined-kotlin-methods)

Jacoco could be appeased by making the `provideParsed` function accept a class type as an argument instead, but that is less idiomatic in my opinion, and leads to unpleasant calling code:

```kotlin
ArgParser(args).provideParsed(ParsedArgs::class)
```

File: README.md [hunk 1]
```diff
@@ -360,6 +360,28 @@ And equally easy to create a program that behaves like `cp`:
...
+  ```
+  
+Alternatively forcing of parsing and validation can be done inline:
+
+  ```kotlin
+  val parsedArgs = ArgParser(args).provideParsed<ParsedArgs>()
+  // now you can use parsedArgs
+  ```
+  
+In both cases exceptions will be thrown where paring or validation errors are found.
```
Comment (xenomachina) [inline README.md]: s/paring/parsing

Comment (shanethehat) [reply README.md]: Oops!

File: README.md [hunk 2]
```diff
@@ -360,6 +360,28 @@ And equally easy to create a program that behaves like `cp`:
   ```
 
 
+## Forcing Parsing
```
Comment (xenomachina) [inline README.md]: I love that you not only documented the new feature, but also improved the documentation for the existing functionality.

File: src/main/kotlin/com/xenomachina/argparser/ArgParser.kt [hunk 3]
```diff
@@ -458,6 +458,15 @@ class ArgParser(args: Array<out String>,
         finished = true
     }
 
+    /**
+     * Provides an instance of T, where all arguments have already been parsed and validated
+     */
+    inline fun <reified T> provideParsed(): T {
```
Comment (xenomachina) [inline src/main/kotlin/com/xenomachina/argparser/ArgParser.kt]: I'm not a big fan of using `getDeclaredConstructor` or `newInstance`, to be honest. I'd rather take a function than a class, as it's both more flexible, and more type safe. (This implementation fails at runtime with a `NoSuchMethodException` if given a class lacking the right type of constructor.)

What do you think about changing it to something like this?:
```
inline fun <T> provideParsed(constructor: (ArgParser) -> T): T {
    val provided = constructor(this)
    force()
    return provided
}
```

This changes the way you call it slightly:
```
parser.provideParsed<Args>()
```
becomes:
```
parser.provideParsed(::Args)
```
but it also lets you pass additional parameters to the Args constructor, or potentially even choose which args class to use based on things like environment variables.
```
parser.provideParsed { parser -> /* code that creates Args */ }
```

Perhaps the name "provideParsed" doesn't work for this different signature. Maybe `parse` or `parseInto` work better with that signature.

I think it'd also make sense to restrict this method to only be used on a "clean" parser. You can add something like this to the end of `ArgParser`:
```
val builtinDelegateCount = delegates.size
```
And then at the beginning of this method add something like:
```
if (delegates.size > builtinDelegateCount) {
    throw IllegalStateException("parseSubCommand can only be used with a clean ArgParser")
}
```

Interestingly, I'm adding a similar method in a branch where I'm working on adding sub-command support. Right now that method's signature looks like:

```
fun <C, S> parseSubCommand(
            commonOptionsConstructor: (OptionParser) -> C,
            subCommandMap: Map<String, (SubCommandArgs<C>) -> S>) : S
```

I'll change the name of that method to be consistent with this one.

Comment (shanethehat) [reply src/main/kotlin/com/xenomachina/argparser/ArgParser.kt]: I like the name `parseInto`. I'll make these changes in the next day or so.

**Comment by xenomachina** [comment]
Hello. Have you had a chance to finish this up? 🙂

File: src/main/kotlin/com/xenomachina/argparser/ArgParser.kt (hunk 3 above)
Comment (xenomachina) [reply src/main/kotlin/com/xenomachina/argparser/ArgParser.kt]: Hello. Have you had a chance to finish this up? 🙂

Comment (shanethehat) [reply src/main/kotlin/com/xenomachina/argparser/ArgParser.kt]: Hey, sorry, I got really busy at work. It's pretty much done, just fighting a failing test.

Comment (shanethehat) [reply src/main/kotlin/com/xenomachina/argparser/ArgParser.kt]: @xenomachina Pushed the requested changes, finally :)

**Pull request merged**